### PR TITLE
[FLINK-19064][hbase] HBaseRowDataInputFormat is leaking resources

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Method `configure()` of `HBaseRowInputFormat` and `HBaseRowDataInputFormat` is calling `connectToTable()`, which creates a connection to HBase that is not closed again. The HBase connection should be opened in `open()` and closed in `close()`.*

## Brief change log

  - *Define `connection` which is opened in `open()` and closed in `close()` in `AbstractTableInputFormat`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by leonardBang — COMMENTED** [review]
Thanks @SteNicholas for the contribution.
One minor comment could we refactor the initialization logic to `OpenInputFormat()` just like `JdbcRowDataInputFormat` did ? I think this also is  the point that @rmetzger and @wuchong  talked under the issue.

**Comment by SteNicholas** [comment]
@leonardBang I have confusion about using ``openInputFormat()`` to refactor. In current logic, ``createInputSplits`` is called before the ``openInputFormat`` method. If we creates a connection to HBase in ``openInputFormat`` method, ``createInputSplits`` causes Exception. Based on above consideration, I don't put the initialization logic to ``openInputFormat()``.

**Comment by wuchong** [comment]
Hi @SteNicholas , I can see your points. However, this can't fix the leaking resources mentioned in the JIRA issue, because the user is using ` DataSourceNode.computeOperatorSpecificDefaultEstimates` which never calls `close()` or `closeInputFormat`. 

We may need to go back to the issue for further discussion.

**Comment by SteNicholas** [comment]
@wuchong I have already changed the behaviour of the `configure()` and `open()` method. The HBase connection is opened in `open()` and closed in `close()`. Could you please review this change request?

**Review by leonardBang — COMMENTED** [review]
Thanks for the update @SteNicholas ,  Could you add a test to cover the `open()` and `close()` methods?

**Comment by SteNicholas** [comment]
@leonardBang I have already add test case for `HBaseTableInputFormat` to verify `open` and `close` method for the HBase connection.

File: flink-connectors/flink-connector-hbase/src/test/java/org/apache/flink/connector/hbase/HBaseTableInputFormatTest.java
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase/src/test/java/org/apache/flink/connector/hbase/HBaseTableInputFormatTest.java:38]: 1.`HBaseTestBase` will pull up a HBase cluster which is a heavy action.
we can move this test to `HBaseConnectorITCase` avoid pull up the cluster twice.

File: flink-connectors/flink-connector-hbase/src/test/java/org/apache/flink/connector/hbase/HBaseTableInputFormatTest.java
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase/src/test/java/org/apache/flink/connector/hbase/HBaseTableInputFormatTest.java:48]: 2.`HBaseRowInputFormat` is used by legacy connector, `HBaseRowDataInputFormat` is used by new connector,
we can cover them in `HBaseConnectorITCase` via parameterized parameter 'isLegacyConnector' easily

**Comment by leonardBang** [comment]
Thanks @SteNicholas, LGTM once the pipeline passed.
Before merge this, I want to CC @rmetzger @wuchong have a look, they gave guidance in the discussion.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/TableInputSplit.java
Comment (wuchong) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/TableInputSplit.java:29]: We don't need to declare it `public`.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (wuchong) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:211]: This is never closed ?

Comment (SteNicholas) [reply flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:211]: > This is never closed ?

This is closed when calling `close` method. I thought that in the lifecycle, `createInputSplits` would be called before `close`. Therefore, I didn't close the connection and table.

Comment (wuchong) [reply flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:211]: I think it is called on the JobMaster [1], so we should use a local connection and close the connection in the method. 

[1]: https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionJobVertex.java#L258

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (rmetzger) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:125]: I believe this (and below) exception messages are not correct anymore.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (rmetzger) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:214]: Revisit exception message.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (rmetzger) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:201]: We will leak the connection if table.close() or resultScanner.close() fails with an exception.
I believe the proper way would be to wrap each close in a separate try block, and catch and log the error.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (rmetzger) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:77]: If you make this method `throws IOException`, you don't need to wrap the IOExceptions in HBaseRowDataInputFormat.connectToTable() into RuntimeExceptions.

Throwing an IOException is not a problem, because initTable is called in open(); which throws an IOException as well.

**Review by rmetzger — COMMENTED** [review]
I left some comments on the code.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (SteNicholas) [reply flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:211]: @wuchong I thought that table isn't close in this method, therefore the connection should not be closed in this method.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/TableInputSplit.java
Comment (SteNicholas) [reply flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/TableInputSplit.java]: @wuchong I have already removed this declaration.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (SteNicholas) [reply flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java]: @rmetzger Good idea. I have followed up with your point to `throws IOException` for `initTable()`.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (SteNicholas) [reply flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:196]: @rmetzger You are right, but `resultScanner.close()` doesn't throw exception. Therefore, I just add the close of the connection into catch block.

**Comment by SteNicholas** [comment]
@rmetzger @wuchong I have already follow up with your comments. Please help to review the changes. Thanks.

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (wuchong) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:192]: nit:

```suggestion
		} finally {
```

File: flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java
Comment (wuchong) [inline flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/AbstractTableInputFormat.java:209]: The duplicate close looks strange to me. I think @rmetzger gave a nice idea above. You can just catch and log the error and move each close in a separate try block. You can take `org.apache.flink.connector.hbase.sink.HBaseSinkFunction#close` as an example. That will be much clean.

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by SteNicholas** [comment]
@rmetzger Could you please take a look at the changes? Any comment?

**Pull request merged**
